src/providers/bibtexformater.ts

```diff
@@ -67,56 +96,71 @@ export class BibtexFormater {
                         entryLocations[i].end.character
                     )
-                    diags.push(new vscode.Diagnostic(
+                    this.diags.push(new vscode.Diagnostic(
                         highlightRange,
                         `Duplicate entry "${entries[i].internalKey}".`,
                         vscode.DiagnosticSeverity.Warning
                     ))
                 } else { // 'Comment Duplicates'
                     // Log duplicate entry since we aren't highlighting it
                     this.extension.logger.addLogMessage(
-                        `BibTeX-format: Duplicate entry "${entries[i].internalKey}" at line ${entryLocations[i].start.line + lineDelta + 1}.`)
+                        `BibTeX-format: Duplicate entry "${entries[i].internalKey}" at line ${entryLocations[i].start.line + lineDelta + 1 + lineOffset}.`)
                     text = text.replace(/@/,'')
                 }
             }
 
             // Put text from entry[i] into (sorted)location[i]
-            edit.replace(uri, entryLocations[i], text)
+            edits.push(new vscode.TextEdit(new vscode.Range(entryLocations[i].start.translate(range?.start.line, range?.start.character), entryLocations[i].end.translate(range?.start.line)), text))
 
             // We need to figure out the line changes in order to highlight properly
             lineDelta += (sortedEntryLocations[i].end.line - sortedEntryLocations[i].start.line) - (entryLocations[i].end.line - entryLocations[i].start.line)
         }
+        this.extension.logger.addLogMessage('Formatted ' + document.fileName)
+        return edits
+    }
+}
 
-        vscode.workspace.applyEdit(edit).then(success => {
-            if (success) {
-                this.duplicatesDiagnostics.set(uri, diags)
-                const t1 = performance.now()
-                this.extension.logger.addLogMessage(`BibTeX action successful. Took ${t1 - t0} ms.`)
-            } else {
-                this.extension.logger.showErrorMessage('Something went wrong while processing the bibliography.')
-            }
-        })
+export class BibtexFormatterProvider implements vscode.DocumentFormattingEditProvider, vscode.DocumentRangeFormattingEditProvider {
+    private formatter: BibtexFormatter
+    extension: Extension
+
+    constructor(extension: Extension) {
+        this.extension = extension
+        this.formatter = new BibtexFormatter(extension)
+    }
+
+    public provideDocumentFormattingEdits(document: vscode.TextDocument, _options: vscode.FormattingOptions, _token: vscode.CancellationToken): vscode.ProviderResult<vscode.TextEdit[]> {
+        const sort = vscode.workspace.getConfiguration('latex-workshop').get('bibtex-format.sort.enabled') as boolean
+        this.extension.logger.addLogMessage('Start bibtex formatting on behalf of VSCode\'s formatter.')
+        return this.formatter.formatDocument(document, sort, true)
```

Comment:
We should catch errors of formatting and add them to log messages, especially for parsing errors.

Reply:
This is resolved by 27c50b1

https://github.com/James-Yu/LaTeX-Workshop/blob/27c50b140ab32b1c1d28dead0eb3311e869ed6a1/src/providers/bibtexformater.ts#L67-L77